### PR TITLE
Basic changes from Dustin's MVP work

### PR DESCRIPTION
**Pull request by dstndstn**
This PR is trying to extract the changes to some function signatures and structures from Dustin's MVP development branch (dstn/frb-bf).  This was done by generating a giant "git diff" and then selectively applying changes with "git commit -p", trying to group commits in a sensible way.

Important changes are:
- Cuda: Create a pipeline state object (cudaPipelineState) that is passed to each stage's execute() function, as a way of passing information between stages
- silly, for debugging: GPU: pass a GPU frame counter that increments each time queue_commands() is called.
- graphviz: record GPU buffers used by each stage, and add them to the graphviz call graph
- Cuda stages: add the option to require a flag to be set in the cudaPipelineState object in order for a stage to run
- GPU memory buffers: add the ability to create GPU memory views, allowing one stage to see only a subset of a memory buffer created by a different stage.  Used in a few places in the MVP to handle rechunking sorts of operations.
- add some float16_t data type handling

And other changes that were hard to extricate from these diffs include:
- cudaBasebandBeamformer: make the 'info' buffer a local buffer (not known to kotekan)

### CODE REVIEW

**Comment by dstndstn** [comment]
btw the Cuda build works

**Comment by dstndstn** [comment]
This plus #1108 make #1078 obsolete

**Comment by dstndstn** [comment]
After switching the macos build from OS version 10.15 to 11, the h5py compile fails;
https://github.com/kotekan/kotekan/actions/runs/5856283980/job/15875669797?pr=1107

**Comment by dstndstn** [comment]
After merging in Jim's CI PR, I dropped the macos-11 build.  Can reinstate it if anyone cares...

**Comment by dstndstn** [comment]
@jbmertens After merging in `develop` to pick up your CI changes, it looks like the 1804 test is failing, but in a way that makes me think this could be an "unlucky" numerical thing -- some of the tests are not deterministic, unfortunately....

**Review by jbmertens — APPROVED** [review]

**Comment by jbmertens** [comment]
I've approved pending lint changes. IWYU is passing, but might not actually be running, which I'll look into in #1117.

**Review by jbmertens — APPROVED** [review]

**Comment by dstndstn** [comment]
Note that all the github-actions commits are irrelevant -- we deleted the macos-11 build while merging in Jim's CI changes on the `develop` branch.

File: lib/cuda/cudaCommand.hpp
Comment (andrerenard) [inline lib/cuda/cudaCommand.hpp:28]: Normally I would say this needs to be documented...  However I think can remove this in the medium term.

File: lib/gpu/gpuProcess.cpp
Comment (andrerenard) [inline lib/gpu/gpuProcess.cpp:171]: Do we want to replace this with a mod counter like `FrameID` now, or address that in another PR?

File: lib/stages/printSparse.hpp
Comment (andrerenard) [inline lib/stages/printSparse.hpp]: This feels like it should be in a `utils/` file, as I don't think this is the only place we are likely to use it.   Also maybe add some simple documentation to explain what the idea of each of these functions are?

File: lib/testing/testDataCheck.hpp
Comment (andrerenard) [inline lib/testing/testDataCheck.hpp]: And it is reused here, so should be in a utils file.

File: tests/boost/CMakeLists.txt
Comment (andrerenard) [inline tests/boost/CMakeLists.txt:68]: Will this build if `cuda` isn't enabled/installed, but boost tests are?   Or are we making cuda a requirement of the boost tests?

**Review by andrerenard — CHANGES_REQUESTED** [review]
This looks good!  I think there are some larger restructuring things to deal with, but not worth holding up this PR.

File: lib/gpu/gpuProcess.cpp
Comment (dstndstn) [reply lib/gpu/gpuProcess.cpp:171]: I'd like to hold off on this one, since I'm hoping for more sweeping changes to this in the near future!

File: lib/testing/testDataCheck.hpp
Comment (dstndstn) [reply lib/testing/testDataCheck.hpp]: Done!

File: lib/stages/printSparse.hpp
Comment (dstndstn) [reply lib/stages/printSparse.hpp]: Done!

File: tests/boost/CMakeLists.txt
Comment (dstndstn) [reply tests/boost/CMakeLists.txt:68]: I just tested with `cmake -DUSE_CUDA=OFF ..` and these tests are still built and they run.  So I think it's fine

File: lib/cuda/cudaCommand.hpp
Comment (dstndstn) [reply lib/cuda/cudaCommand.hpp:28]: hah!, sweet, I got away with it!

**Comment by dstndstn** [comment]
Hallelujah!!
Thanks!!

**Pull request merged**
